### PR TITLE
deduce the second argument to iconv()

### PR DESCRIPTION
**Pull request by arvidn**
instead of requiring it to be configured via a macro

### CODE REVIEW

**Comment by arvidn** [comment]
it seems GCC didn't fix ``decltype(function)`` until 6.1, which might be a bit of a strict compiler requirement.

**Comment by fmatthew5876** [comment]
Nice.

That's way better than the #defines. The only suggestion I would make is to wrap the `call_iconv()` with an inline `libtorrent::iconv()` function one more time so clients don't need to understand how the function pointer trick works.

**Comment by arvidn** [comment]
the only place ``iconv()`` is called from is ``iconv_convert_impl()`` afaict.

Now I just need to figure out how to suppress the unused function warning in gcc.

**Comment by arvidn** [comment]
@fmatthew5876 ok, I think this patch is in pretty good shape now. Would you mind testing it to make sure it fixes the original issue you experienced?

File: src/escape_string.cpp [hunk 1]
```diff
@@ -523,11 +534,10 @@ namespace {
 		ret.resize(outsize);
 		char const* in = s.c_str();
 		char* out = &ret[0];
-		// posix has a weird iconv signature. implementations
-		// differ on what this signature should be, so we use
-		// a macro to let config.hpp determine it
-		size_t retval = iconv(h, TORRENT_ICONV_ARG(&in), &insize,
-			&out, &outsize);
+		// posix has a weird iconv() signature. implementations
+		// differ on the trype of the second parameter. We use a helper template
```
Comment (aldenml) [inline src/escape_string.cpp]: no important typo `trype`

**Comment by fmatthew5876** [comment]
Works for me

**Pull request merged**